domains/client/domain-operator/src/snap_sync.rs

```diff
@@ -0,0 +1,478 @@
+    )
+    .await;
+
+    trace!("State downloaded: {:?}", state_result);
+
+    {
+        let client = sync_params.domain_client.clone();
+        // Import first block as finalized
+        let mut block =
+            BlockImportParams::new(BlockOrigin::NetworkInitialSync, domain_block_header);
+        block.body = domain_block.body;
+        block.justifications = domain_block.justifications;
+        block.state_action = StateAction::ApplyChanges(StorageChanges::Import(state_result?));
+        block.finalized = true;
+        block.fork_choice = Some(ForkChoiceStrategy::Custom(true));
+        client.as_ref().import_block(block).await.map_err(|error| {
+            sp_blockchain::Error::Backend(format!("Failed to import state block: {error}"))
+        })?;
+    }
+
+    wait_for_block_import(
+        sync_params.domain_client.as_ref(),
+        domain_block_number.into(),
+    )
+    .instrument(info_span!("domain chain snap sync"))
+    .await;
+
+    trace!(
+        "Domain client info after waiting: {:?}",
+        sync_params.domain_client.info()
+    );
+
+    // Verify domain state block creation.
+    if let Ok(Some(created_domain_block_hash)) =
+        sync_params.domain_client.hash(domain_block_number.into())
+    {
+        if created_domain_block_hash == domain_block_hash {
+            trace!(
+                ?created_domain_block_hash,
+                ?domain_block_hash,
+                "Created hash matches after the domain block import with state",
+            );
+        } else {
+            debug!(
+                ?created_domain_block_hash,
+                ?domain_block_hash,
+                "Created hash doesn't match after the domain block import with state",
+            );
+
+            return Err(sp_blockchain::Error::Backend(
+                "Created hash doesn't match after the domain block import with state".to_string(),
+            ));
+        }
+    } else {
+        return Err(sp_blockchain::Error::Backend(
+            "Can't obtain domain block hash after state importing for snap sync".to_string(),
+        ));
+    }
+
+    crate::aux_schema::track_domain_hash_and_consensus_hash(
+        sync_params.domain_client.as_ref(),
+        domain_block_hash,
+        consensus_block_hash,
+    )?;
+
+    crate::aux_schema::write_execution_receipt::<_, Block, CBlock>(
+        sync_params.domain_client.as_ref(),
+        None,
+        &last_confirmed_block_receipt,
+    )?;
+
+    if let Some(offchain_storage) = sync_params
+        .consensus_chain_sync_params
+        .backend
+        .offchain_storage()
+    {
+        let target_block = sync_params
+            .consensus_chain_sync_params
+            .segment_headers_store
+            .max_segment_index()
+            // Skip last `FINALIZATION_DEPTH_IN_SEGMENTS` archived segments
+            .and_then(|max_segment_index| {
+                max_segment_index.checked_sub(FINALIZATION_DEPTH_IN_SEGMENTS)
+            })
```

Comment:
We discussed that it is not actually needed to download older segment at `FINALIZATION_DEPTH_IN_SEGMENTS` because it is equivalent to downloading the latest segment if responder can do a little bit of custom logic composing necessary data from technically not yet "finalized" from Substrate's point of view blocks. What happened to that?

Reply:
This is my current task and I will issue a separate PR for that similar to other MMR-sync updates.,